droid-results/src/main/java/uk/gov/nationalarchives/droid/results/handlers/WriterResultHandlerDao.java

```diff
@@ -0,0 +1,177 @@
+import uk.gov.nationalarchives.droid.profile.ProfileResourceNodeFilter;
+
+
+/**
+ * An implementation of ResultHandlerDao which writes results out to a Writer,
+ * formatting them using an ItemWriter, instead of out to the profile database.
+ * This can be used to output results to CSV on the console for example, or XML to a file.
+ * <p>
+ * We inherit from the JDBCBatchResultHandlerDao class in order to get its
+ * database operations for initialising the db and loading lists of formats.
+ * Although these functions probably shouldn't belong in result handlers,
+ * it's where they are now, and refactoring this turns out to be fairly
+ * complex.  So we re-use all the database parts from the JDBCBatchResultHandler,
+ * but override the methods involved in writing out to a Writer, and those
+ * which can't work if the results aren't being saved to the database
+ * (e.g. loading previous results or deleting previous results).
+ */
+public class WriterResultHandlerDao extends JDBCBatchResultHandlerDao {
+
+    private static final Logger LOG = LoggerFactory.getLogger(WriterResultHandlerDao.class);
+
+    private final List<ProfileResourceNode> items = new ArrayList<>(1);
+
+    private ItemWriter itemWriter;
+    private Writer writer;
+    private long nodeId = 1L;
+    private ProfileResourceNodeFilter filter = new ProfileResourceNodeFilter();
+
+    /**
+     * Empty bean constructor.  You still need to set the Itemwriter and DataSource,
+     * and then call init() before this class is ready to use.  If you don't also set a Writer,
+     * then output will go to the standard system out.
+     */
+    public WriterResultHandlerDao() {
+    }
+
+    /**
+     * Parameterized constructor taking the ItemWriter to use to output results to standard out.
+     * @param itemWriter The ItemWriter to write out results.
+     * @param datasource The database connection to the profile for initialising new databases.
+     */
+    public WriterResultHandlerDao(ItemWriter itemWriter, DataSource datasource) {
+        this(itemWriter, null, datasource);
+    }
+
+    /**
+     * Parameterized constructor taking the ItemWriter to use to output results to a Writer.
+     * @param itemWriter The ItemWriter which formats results and writes them to the writer.
+     * @param writer The writer to which results are written.
+     * @param datasource The database connection to the profile for initialising new databases.
+     */
+    public WriterResultHandlerDao(ItemWriter itemWriter, Writer writer, DataSource datasource) {
+        setItemWriter(itemWriter);
+        setWriter(writer);
+        setDatasource(datasource);
+    }
+
+    @Override
+    public synchronized void init() {
+        try {
+            super.init();
+        } finally {
+            if (writer == null) { // If no writer is set, default to console output.
+                writer = new PrintWriter(System.out);
+            }
+            itemWriter.open(writer);
```

Comment:
Do we not need to close this itemWriter

Reply:
Would be better to close it I think, good catch.  There's no close() method in any of the results handling interfaces, so would need to add those and wire it all up.  I'll have a look at doing that.

Reply:
Turns out adding close() is harder than it looks.  Since processing multi-threaded, if you close a profile as soon as you've finished walking the profile resources, there can still be identification requests in flight on other threads.  This can cause the underlying connections (file or database) to close before all identification is finished, throwing various errors and failing to record the last profile results.

Reply:
Would need to have some kind of thread safe signal that everything was finished in order to allow safe close.

Reply:
I may have a look at this - but I don't think this is a major issue.  What would be the worst effect of not closing the CSV writer and underlying file access?  On the command line, execution ceases as soon as the profile is complete, so there won't exactly be any resource leak.  This writer can't be invoked from the GUI (which writes to database).

Reply:
Some investigation says that the primary effect of not explicitly closing something is that there may be unflushed data which may not get written.  The current `CsvItemWriter` code calls `flush()` after each file it writes out so there is no risk currently we will not write all data out.

Reply:
It does make me wonder though what happens when the command line closes down.  If it closes down as soon as all profile resources are finished, what does that do to threads that are in flight?  Could we lose the last few identifications?  Doesn't seem to have been a problem for writing to the database when profiling on the command line, but might bear a bit more investigation.